inspirehep/modules/records/json_ref_loader.py

```diff
@@ -0,0 +1,122 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of INSPIRE.
+# Copyright (C) 2016 CERN.
+#
+# INSPIRE is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License as
+# published by the Free Software Foundation; either version 2 of the
+# License, or (at your option) any later version.
+#
+# INSPIRE is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with INSPIRE; if not, write to the Free Software Foundation, Inc.,
+# 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
+
+"""Resource-aware json reference loaders to be used with jsonref."""
+
+from jsonref import JsonLoader, JsonRef
+from werkzeug.urls import url_parse
+
+from flask import current_app
+
+from invenio_records.api import Record
+from invenio_pidstore.models import PersistentIdentifier
+from invenio_search import current_search_client as es
+
+
+class AbstractRecordLoader(JsonLoader):
+    """Base for resource-aware record loaders.
+
+    Resolves the refered resource by the given uri by first checking against
+    local resources.
+    """
+
+    def get_record(self, record_type, recid):
+        raise NotImplementedError()
+
+    def get_remote_json(self, uri, **kwargs):
+        parsed_uri = url_parse(uri)
+        # Normalize optional 'http://' protocol in the config.
+        server_name = current_app.config.get('SERVER_NAME')
+        if not server_name.startswith('http://'):
+            server_name = 'http://{}'.format(server_name)
+        parsed_server = url_parse(server_name)
+
+        if parsed_uri.netloc and parsed_uri.netloc != parsed_server.netloc:
+            return super(AbstractRecordLoader, self).get_remote_json(uri,
+                                                                     **kwargs)
+        path_parts = parsed_uri.path.strip('/').split('/')
+        if len(path_parts) < 2:
+            current_app.logger.error('Bad JSONref URI: {0}'.format(uri))
+            return None
+
+        res = None
+        try:
+            record_type = path_parts[-2]
+            recid = path_parts[-1]
+            res = self.get_record(record_type, recid)
+        except Exception as e:
+            current_app.logger.error('Encountered Exception {0} while '
+                                     'retrieving record [{1}:{2}]'.format(
+                                         e, record_type, recid))
+        finally:
+            return res
+
+
+class ESJsonLoader(AbstractRecordLoader):
+    """Resolve resources by retrieving them from Elasticsearch."""
+
+    def get_record(self, record_type, recid):
```

Comment:
I am not the best in suggesting how to structure our overlay. @jalavik, @jmartinm what do you think? I guess you can put this `record_getter.py` file simply under utils.